### PR TITLE
fix: potential static caching memory exhaustion

### PR DESCRIPTION
**Pull request by davwheat**
**Changes proposed in this pull request:**

Removes static caching and model filtering in favour of adding extra where clauses to a relation.

**Reviewers should focus on:**

Why was the static caching utilised in the first place?

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

File: framework/core/src/User/User.php [hunk 1]
```diff
@@ -439,7 +439,21 @@ public function getAlertableNotificationTypes()
...
-        return $this->getUnreadNotifications()->count();
+        return $this->queryUnreadNotifications()->count();
+    }
+
+    /**
+     * Return query builder for all notifications that have not been read yet.
+     *
+     * @return \Illuminate\Database\Eloquent\Relations\HasMany
+     */
+    protected function queryUnreadNotifications()
```
Comment (davwheat) [inline framework/core/src/User/User.php]: should probably just be `unreadNotifications()`

**Review by SychO9 — COMMENTED** [review]
This seems to be the only relevant discussion: https://github.com/flarum/framework/issues/500. As far as I can tell the static caching was only done for the sake of avoiding two counting queries. But I'd rather have two count queries than do this caching and risk causing bugs/memory exhaustion issues as we saw in realtime ext.

So I'm in favor of moving forward with this.

File: framework/core/src/User/User.php [hunk 2]
```diff
@@ -470,9 +473,9 @@ protected function getUnreadNotifications()
      */
     public function getNewNotificationCount()
     {
-        return $this->getUnreadNotifications()->filter(function ($notification) {
-            return $notification->created_at > $this->read_notifications_at ?: 0;
-        })->count();
+        return $this->queryUnreadNotifications()
```
Comment (davwheat) [inline framework/core/src/User/User.php]:
```suggestion
        return $this->unreadNotifications()
```

File: framework/core/src/User/User.php [hunk 3]
```diff
@@ -449,18 +463,7 @@ public function getUnreadNotificationCount()
...
-            $cached[$this->id] = $this->notifications()
-                ->whereIn('type', $this->getAlertableNotificationTypes())
-                ->whereNull('read_at')
-                ->where('is_deleted', false)
-                ->whereSubjectVisibleTo($this)
-                ->get();
-        }
-
-        return $cached[$this->id];
+        return $this->queryUnreadNotifications()->get();
```
Comment (davwheat) [inline framework/core/src/User/User.php]:
```suggestion
        return $this->unreadNotifications()->get();
```

File: framework/core/src/User/User.php [hunk 4]
```diff
@@ -439,7 +439,21 @@ public function getAlertableNotificationTypes()
      */
     public function getUnreadNotificationCount()
     {
-        return $this->getUnreadNotifications()->count();
+        return $this->queryUnreadNotifications()->count();
```
Comment (davwheat) [inline framework/core/src/User/User.php]:
```suggestion
        return $this->unreadNotifications()->count();
```

File: framework/core/src/User/User.php (hunk 1 above)
Comment (davwheat) [reply framework/core/src/User/User.php]:
```suggestion
    protected function unreadNotifications()
```

**Review by luceos — APPROVED** [review]

**Pull request merged**